### PR TITLE
Centralize FastForwardBus scheduler, add TestDriver.assertWith, refactor VM tests

### PR DESCRIPTION
**Pull request by ducky64**
It's all related, I swear!

Builds on top of #239 and #237, probably the last in a series of testing PRs.

- The prior `FastForwardBus` only allowed one `runToPromise` / `runForDelay` at any instant, which limits expressiveness where concurrent testdriver operations may be desired (for example, overlapping test stimulus with assertion conditions). This centralizes the scheduler, which now advances time as long as there is some testdriver operation (`runToPromise`, `runForDelay`) pending.
- Adds a `FastForwardTester.assertWith(events, body)`, which runs a `body` function containing test stimulus (eg, button down, wait, button up, wait) but requires the `events` to trigger (or hold, as applicable) during `body` execution. Might be useful to back-port to the base `TestDriver` eventually, but overlapping test stimulus probably isn't useful for a physical device.
- Refactors VM tests to use the above constructs.

Minor bugfixes and refactorings:
- Fix edge condition with `RegisterTester.hold(...)` not accepting the exact specified value
- Add `ServiceTester.hold()` as a condition that asserts no events are raised

Overall, I think this expresses what I would like to express in the test cases, and has good coverage in time (as opposed to assertions that only check an instant in time). I'm not totally happy with how the test cases look - might be a side effect of prettier - but not sure what can be done there.

### CODE REVIEW

File: src/tstester/servicewrapper.ts [hunk 1]
```diff
@@ -22,13 +19,21 @@ class BaseServiceEventTrigger extends EventWithHoldAdapter<JDEvent> {
...
+export class ServiceHold extends BaseServiceEventTrigger {
+    protected processHold(data: JDEvent) {
+        throw new Error(
+            `service ${this.service.name} got event ${data.code} (${data.name}) when hold asserted`
+        )
+    }
+}
+
 // Event that fires on a matching event code from the specified service
 class BaseServiceAnyEventTrigger extends BaseServiceEventTrigger {
```
Comment (pelikhan) [inline src/tstester/servicewrapper.ts:31]: same comment about base

File: tests/jdom/fastforwardtester.ts [hunk 2]
```diff
@@ -84,10 +85,12 @@ export class FastForwardTester
     // start() and stop() are made available, but recommend using withTestBus() instead
     public start() {
         this.bus.start()
+        this.scheduler.start() // run in background
```
Comment (pelikhan) [inline tests/jdom/fastforwardtester.ts:88]: isn't the bus responsability to start/stop the scheduler?

File: tests/jdom/fastforwardtester.ts [hunk 3]
```diff
@@ -84,10 +85,12 @@ export class FastForwardTester
...
     }
 
     public stop() {
         this.bus.stop()
+        this.scheduler.stop()
```
Comment (pelikhan) [inline tests/jdom/fastforwardtester.ts:93]: same comment

File: tests/vm/buttonservo.spec.ts [hunk 4]
```diff
@@ -37,45 +39,78 @@ suite("button servo", () => {
...
-            button.server.down()
-            await tester.waitForDelay(100)
-            button.server.up()
-            assert(servo.server.angle.values()[0] == -50.0)
+            await tester.assertWith(
+                servoReg
+                    .onUpdate({
+                        triggerRange: [-50, -50],
+                    })
+                    .hold(),
```
Comment (pelikhan) [inline tests/vm/buttonservo.spec.ts:59]: i have a hard time grasping what hold is supposed to mean. can we use a more explicit (potentially longer) API name?

File: tests/vm/buttonservo.spec.ts [hunk 5]
```diff
@@ -37,45 +39,78 @@ suite("button servo", () => {
...
     test(
         "does not invert when not pressed",
-        makeVmTest(async (tester, button, servo) => {
-            servo.server.angle.setValues([50])
+        makeVmTest(async (tester, button, servo, servoReg) => {
+            servo.angle.setValues([50])
 
-            await tester.waitForDelay(100)
-            assert(servo.server.angle.values()[0] == 50)
+            await tester.assertWith(servoReg.hold([50, 50]), async () => {
```
Comment (pelikhan) [inline tests/vm/buttonservo.spec.ts:75]: what is the meaning of ``[50,50]``?

**Review by pelikhan — APPROVED** [review]

**Review by tballmsft — APPROVED** [review]

File: tests/vm/buttonservo.spec.ts (hunk 5 above)
Comment (ducky64) [reply tests/vm/buttonservo.spec.ts:75]: [50, 50] is meant as a range, but I think it will probably be a better API to allow it to take either a single number (exact match) in addition to the interval it currently supports.

File: tests/vm/buttonservo.spec.ts (hunk 4 above)
Comment (ducky64) [reply tests/vm/buttonservo.spec.ts:59]: Yeah, fair point. The idea is to have it trigger when the value is within some range, but it also supports triggering on any update (which isn't used).

It probably makes more sense to nix the second and name it so it's specific to the first - perhaps `onValue` - and require the triggerRange as the first argument while moving other conditions to a secondary options argument.

File: tests/jdom/fastforwardtester.ts (hunk 3 above)
Comment (ducky64) [reply tests/jdom/fastforwardtester.ts:93]: Yeah, so the issue is that there isn't a `start`/`stop` interface for the base `Scheduler`, and `stop` doesn't really make sense for the wall clock scheduler.

Potential options include:
- keep as-is - which might be the least worst option, since FastForwardScheduler is tightly integrated with the FastForwardTester anyways
- have the FastForwardScheduler automatically start on construction, but keep the explicit `stop` so it can be removed from the event queue
- add a start/stop interface on the schedulers, and have the WallClockScheduler no-op when asked to `stop` - which seems like a poor API design
- actually have the WallClockScheduler be `stop`pable - which is clean, but doesn't seem to be useful
- add an intermediate class of StartStopScheduler, which FastForwardScheduler would extend (but WallClockScheduler would not), and which the bus is aware of and can forward `start` and `stop` calls

File: src/tstester/servicewrapper.ts (hunk 1 above)
Comment (ducky64) [reply src/tstester/servicewrapper.ts:32]: Not completely sure what you're referring to, but added and updated some comments.
